core/src/main/java/org/apache/calcite/rex/RexSimplify.java

```diff
@@ -708,12 +708,15 @@ private RexNode simplifyIs2(SqlKind kind, RexNode a, RexUnknownAs unknownAs) {
   }
 
   private RexNode simplifyIsNotNull(RexNode a) {
-    if (!a.getType().isNullable()) {
+    if (!a.getType().isNullable() && isSafeExpression(a)) {
       return rexBuilder.makeLiteral(true);
     }
     if (predicates.pulledUpPredicates.contains(a)) {
```

Comment:
Would this suffer from the same "not that safe expression" problem?

Reply:
I think if its a pulledup predicate - then that means it is already evaluated somewhere else as well

Reply:
I saw this conversation and assumed it wasn't blocking the merge of this change.

@vlsi / @kgyrtkirk - If I was wrong and this should have held up the merge of this - please ping me. I have started the 1.19.0 RC2 process